tests/CustomMatchers/CustomMatchersForRendering.ts

```diff
@@ -17,10 +17,16 @@ declare global {
 }
 
 function getDataAttributesAsString(element: HTMLElement): string {
-    const dataAttributes = element.dataset;
-    const keys = Object.keys(dataAttributes);
+    const outerHTML = element.outerHTML;
 
-    return keys.map((key) => `${key}: ${dataAttributes[key]}`).join('\n');
+    const dataAttributesRegex = /data-[a-zA-Z0-9-]+="[^"]*"/g;
```

Comment:
I don't know if we've had any conversations about regular expressions ... if we have, I would have said that they invariably hide complexity and errors.

So my first question about this new way of testing is:
'How do we know that this regex is guaranteed to satisfy all possible future values ever?". For example, does HTML, CSS or whatever place some kind or restriction on the valid characters? For example, is underscore guaranteed not to be allowed?

Reply:
And could there ever be an attribute that does not begin with `data-`?

Reply:
With that one commit this PR went from 'OK, cool - I will merge that when I'm back in front of a computer' to ...

Oh right, now we probably need unit tests for the tests, since they are non-trivial to reason about.

And maybe we need checks in the production code to ensure that no code is ever added that breaks the assumptions in the tests...

Reply:
> I don't know if we've had any conversations about regular expressions ... if we have, I would have said that they invariably hide complexity and errors.
> 
> So my first question about this new way of testing is: 'How do we know that this regex is guaranteed to satisfy all possible future values ever?". For example, does HTML, CSS or whatever place some kind or restriction on the valid characters? For example, is underscore guaranteed not to be allowed?

Well, there is this spec https://html.spec.whatwg.org/multipage/dom.html#embedding-custom-non-visible-data-with-the-data-*-attributes that describes the data attributes and implicitly the underscores are not allowed.

> And could there ever be an attribute that does not begin with `data-`?

No according to the spec above.


Yet your point is still valid. So I will drop the latest commit and force push the thing. Thanks for enlightening me on the regex. They raise a lot of questions indeed.

Reply:
Thank you for the very helpful reply.
That's a really useful link.

Reply:
I've dropped the commit introducing the regex and pushed the changes =)

Reply:
For future reference...
Having looked more closely that the link above about data attributes, I'm wondering why a regular expression was needed at all...
Would `attribute.startsWith('data-')` have done the job?

Reply:
Well the structrure of the HTML there is  `<tag non-data-attribute="value1" data-attribute="value2">`. So there is a need to filter the `<tag` and the `>` at the end, which is not hard but makes the code complex to understand (same thing with regex). Think this is already an improvement (having a matcher with a clear statement), maybe we can make it better when we pair. For now I suggest to merge as is.

Reply:
Could do something like this:

```ts
function getDataAttributesAsString(element: HTMLElement): string {
    const attributes = Array.from(element.attributes);
    const dataAttributes = attributes.filter((attribute) => attribute.name.startsWith('data-'));
    return dataAttributes.map((attribute) => `${attribute.name}="${attribute.value}"`).join('\n');
}
```

Which would output attributes as `data-task-priority="highest"`. But the whole scope of getting the attributes from the HTML was to test them in the output (the HTML) and not `HTMLELement` inner data processed in any way (`${attribute.name}="${attribute.value}"`). So if I compare this solution (in this comment) and the one in the current code, I prefer the latter because it is not modifying the data attributes and doesn't give any "false hopes" to someone seeing the test results.

Reply:
To give more context about the data attributes, they are accessible with:
- `HTMLElement.outerHTML` along with the HTML tag, here everything is "as rendered"
  - pros: testing the real HTML output
  - cons: need something to extract the data attributes
- `HTMLElement.attributes` dictionary with key in kebab case (+ need to filter the data attributes from here)
  - pros: none
  - cons: not the real output, need to filter data attributes, difference between the output and what is set in the code (`taskDue` vs `data-task-due`)
- `HTMLElement.dataset` dictionary with keys in camel case - the way we set them in the code (`taskDue`)
  -  pros: simple, same data representation as in the code
  - cons: not the real output

So it also makes sense to

Reply:
... makes sense to what?

Reply:
Oops. So it also makes sense to leave things as they are (testing the `dataset`).

Reply:
Thanks.